File: 152-m-dp/152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxsf = nums.at(0);
        int minsf = nums.at(0);
        int maxans = *std::max_element(nums.begin(),nums.end());

        int t;
        for (auto n = nums.begin() + 1; n != nums.end(); n++){
            if ((*n) == 0){
                maxsf = 1;
                minsf = 1;
                continue;
            }

            t = maxsf * (*n);
            maxsf = std::max((*n), std::max((*n) * minsf, t));
            minsf = std::min((*n), std::min((*n) * minsf, t));

            maxans = (std::max(maxsf, minsf) > maxans) ? std::max(maxsf, minsf) : maxans;
        }

        return maxans;
    }
};

File: 152-m-dp/152.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        maxsf = nums[0]
        minsf = nums[0]
        maxans = max(nums)

        for n in nums[1:]:
            if n == 0:
                maxsf, minsf = 1, 1
                continue

            t = n * maxsf
            maxsf = max(t, n * minsf, n)
            minsf = min(t, n * minsf, n)

            maxans = max(maxsf, minsf, maxans)

        return maxans